File: src/scenes/file_player/grade_text.rs
//! The judgment word that pops on each graded row, shaded live: the word
//! is rendered white to an offscreen image so its alpha is pure coverage,
//! then presented on a quad whose material tints it to the grade color and
//! layers on an additive glow that pulses — one shader, per-grade colors
//! and strengths (see [`grade_text.wgsl`](../../../assets/shaders/grade_text.wgsl)).

use super::{ForPlayer, PlaySet, RowGraded};
use crate::core::config::{DynamicGradeDef, GameConfig, Grade, RowOutcome, TimingFeedback};
use crate::core::font::game_font;
use crate::core::note_field::visible_world_size;
use crate::core::player::PlayerId;
use crate::core::settings::PlayerSettings;
use crate::core::units::{Percent, Seconds};
use crate::scenes::GameScene;
use bevy::camera::visibility::RenderLayers;
use bevy::camera::{ClearColorConfig, RenderTarget, ScalingMode};
use bevy::mesh::MeshVertexBufferLayoutRef;
use bevy::prelude::*;
use bevy::render::render_resource::{
    AsBindGroup, BlendState, RenderPipelineDescriptor, SpecializedMeshPipelineError, TextureFormat,
};
use bevy::shader::ShaderRef;
use bevy::sprite_render::{
    AlphaMode2d, Material2d, Material2dKey, Material2dPlugin, MeshMaterial2d,
};

/// The word's font height, in world units.
const FONT_SIZE: f32 = 50.0;
/// World size of the presented quad; the offscreen camera frames exactly
/// this, wide enough to hold the longest timing-feedback string.
pub const PRESENT_W: f32 = 760.0;
pub const PRESENT_H: f32 = 170.0;
/// The offscreen image is rendered larger than presented so the word stays
/// crisp when the window scales the quad up.
const SUPERSAMPLE: u32 = 2;
const IMAGE_W: u32 = PRESENT_W as u32 * SUPERSAMPLE;
const IMAGE_H: u32 = PRESENT_H as u32 * SUPERSAMPLE;
/// The neon glow's reach around the glyphs, in world units, and its
/// brightness.
const HALO_RADIUS: f32 = 18.0;
const HALO_STRENGTH: f32 = 1.6;
/// The glow strikes to full the instant a grade lands, then drains toward
/// [`GLOW_FLOOR`] with this time constant — a per-hit pulse, not a steady
/// shine, in the spirit of the wheel and health-bar pulses.
const PULSE_TAU: f32 = 0.3;
const GLOW_FLOOR: f32 = 0.32;
/// The grade group's vertical extent, keeping the word and the combo under
/// it inside the screen's padded band.
const GRADE_HALF_HEIGHT: f32 = 36.0;
const COMBO_HALF_HEIGHT: f32 = 24.0;
/// The combo readout sits this far under the grade's center.
pub const COMBO_GAP: f32 = 62.0;
/// The pop each grade opens with: a brief upscale that settles.
const BOUNCE_SECONDS: f32 = 0.13;
const BOUNCE_AMOUNT: f32 = 0.18;
/// Seconds a grade takes to fade out once the player stops hitting.
const FADE_SECONDS: f32 = 1.0;
const GRADE_Z: f32 = 6.0;
/// The offscreen cameras render to images, not the window, so their order
/// only needs to be distinct and out of the window cameras' way.
const SOURCE_CAMERA_ORDER: isize = -100;

pub(super) fn plugin(app: &mut App) {
    app.add_plugins(Material2dPlugin::<GradeTextMaterial>::default())
        .add_systems(
            Update,
            (apply_grades, animate_grades)
                .chain()
                .in_set(PlaySet::Present),
        )
        .add_systems(
            Update,
            set_stage_grade_area
                .in_set(PlaySet::Present)
                .run_if(in_state(GameScene::FilePlayer)),
        );
}

/// The entities of one grade-text rig, for callers to position, re-layer,
/// and scope to their own context.
pub struct GradeRig {
    /// The shader quad presenting the word.
    pub present: Entity,
    /// The offscreen white word; set its `Text2d` to change the grade.
    pub source: Entity,
    /// The offscreen camera rendering `source` into the sampled image.
    pub camera: Entity,
    pub material: Handle<GradeTextMaterial>,
}

/// Builds a grade-text rig drawing on the private `source_layer`: the
/// offscreen white word, its camera and image, and the shader quad
/// presenting it at the origin on the default layer. Callers position,
/// re-layer, and scope the returned entities. Shared by the play stage,
/// the options preview, and the `render_grade` inspector.
pub fn spawn_rig(
    commands: &mut Commands,
    images: &mut Assets<Image>,
    asset_server: &AssetServer,
    source_layer: usize,
) -> GradeRig {
    let layer = RenderLayers::layer(source_layer);
    // Added straight to `Assets` (not deferred through the asset server) so
    // the offscreen camera's target resolves the very frame it spawns.
    let image = images.add(Image::new_target_texture(
        IMAGE_W,
        IMAGE_H,
        TextureFormat::Rgba8UnormSrgb,
        None,
    ));
    let camera = commands
        .spawn_scene(bsn! { Camera2d })
        .insert((
            Camera {
                order: SOURCE_CAMERA_ORDER - source_layer as isize,
                clear_color: ClearColorConfig::Custom(Color::NONE),
                ..default()
            },
            RenderTarget::Image(image.clone().into()),
            Projection::Orthographic(OrthographicProjection {
                scaling_mode: ScalingMode::Fixed {
                    width: PRESENT_W,
                    height: PRESENT_H,
                },
                ..OrthographicProjection::default_2d()
            }),
            layer.clone(),
        ))
        .id();
    let source = commands
        .spawn_scene(bsn! {
            GradeSource
            game_font(FONT_SIZE)
            Text2d("")
            TextColor(Color::WHITE)
        })
        .insert(layer)
        .id();
    let material = asset_server.add(GradeTextMaterial {
        base: Vec4::new(1.0, 1.0, 1.0, 0.0),
        glow: Vec4::ZERO,
        shape: Vec4::new(
            HALO_RADIUS / PRESENT_W,
            HALO_RADIUS / PRESENT_H,
            HALO_STRENGTH,
            0.0,
        ),
        text: image,
    });
    let mesh = asset_server.add(Mesh::from(Rectangle::new(PRESENT_W, PRESENT_H)));
    let present = commands
        .spawn((
            Mesh2d(mesh),
            MeshMaterial2d(material.clone()),
            Transform::from_xyz(0.0, 0.0, GRADE_Z),
        ))
        .id();
    GradeRig {
        present,
        source,
        camera,
        material,
    }
}

/// Where and how a grade-text display is placed: its player, the field's
/// x center, the private layer its offscreen word renders on, and the layer
/// its shader quad presents on (`None` for the default layer).
pub struct GradeSpawn {
    pub player: PlayerId,
    pub origin_x: f32,
    pub source_layer: usize,
    pub present_layer: Option<usize>,
}

/// Spawns a grade-text display driven by the shared [`apply_grades`] /
/// [`animate_grades`] systems: the offscreen rig plus the [`GradeText`]
/// state, scoped to `scope`. Returns the quad entity. Used by the play stage
/// and the options preview alike.
pub fn spawn_display(
    commands: &mut Commands,
    images: &mut Assets<Image>,
    asset_server: &AssetServer,
    spec: GradeSpawn,
    scope: impl Bundle + Clone,
) -> Entity {
    let rig = spawn_rig(commands, images, asset_server, spec.source_layer);
    for entity in [rig.camera, rig.source] {
        commands.entity(entity).insert(scope.clone());
    }
    let mut present = commands.entity(rig.present);
    present.insert((
        GradeText {
            player: spec.player,
            source: rig.source,
            intensity: 0.0,
            pulse: 0.0,
            bounce: 0.0,
            base: Color::WHITE,
            glow: Color::WHITE,
            strength: 0.0,
        },
        ForPlayer(spec.player),
        Transform::from_xyz(spec.origin_x, 0.0, GRADE_Z),
        scope,
    ));
    if let Some(layer) = spec.present_layer {
        present.insert(RenderLayers::layer(layer));
    }
    rig.present
}

/// The offscreen word entity, rendered white so the shader owns its color.
#[derive(Component, Default, Clone)]
struct GradeSource;

/// The presented shader quad and its running animation state.
#[derive(Component)]
struct GradeText {
    player: PlayerId,
    source: Entity,
    /// Fade level: 1 on a fresh grade, decaying to 0 once hits stop.
    intensity: f32,
    /// Seconds since the last grade, driving the glow's oscillation.
    pulse: f32,
    /// The pop-in bounce, 1 on a fresh grade decaying to 0.
    bounce: f32,
    /// The current grade's color, glow color, and glow strength.
    base: Color,
    glow: Color,
    strength: f32,
}

#[derive(AsBindGroup, Asset, TypePath, Clone)]
pub struct GradeTextMaterial {
    /// `(linear grade color, fade)`.
    #[uniform(0)]
    pub base: Vec4,
    /// `(linear glow color, strength times the pulse)`.
    #[uniform(1)]
    pub glow: Vec4,
    /// `(halo radius u, halo radius v, halo strength, unused)`.
    #[uniform(2)]
    pub shape: Vec4,
    #[texture(3)]
    #[sampler(4)]
    pub text: Handle<Image>,
}

impl Material2d for GradeTextMaterial {
    fn fragment_shader() -> ShaderRef {
        "shaders/grade_text.wgsl".into()
    }

    fn alpha_mode(&self) -> AlphaMode2d {
        AlphaMode2d::Blend
    }

    /// The shader outputs premultiplied color, so the glow adds light on top
    /// of any background instead of blending toward it.
    fn specialize(
        descriptor: &mut RenderPipelineDescriptor,
        _layout: &MeshVertexBufferLayoutRef,
        _key: Material2dKey<Self>,
    ) -> Result<(), SpecializedMeshPipelineError> {
        if let Some(fragment) = &mut descriptor.fragment {
            for target in fragment.targets.iter_mut().flatten() {
                target.blend = Some(BlendState::PREMULTIPLIED_ALPHA_BLENDING);
            }
        }
        Ok(())
    }
}

/// Each graded row refreshes its player's word, color, and glow, and
/// restarts the pop and fade.
fn apply_grades(
    config: Res<GameConfig>,
    mut graded: MessageReader<RowGraded>,
    mut grades: Query<&mut GradeText>,
    mut sources: Query<&mut Text2d, With<GradeSource>>,
) {
    for message in graded.read() {
        for mut grade in &mut grades {
            if grade.player != message.player {
                continue;
            }
            let style = grade_style(&config, message.outcome);
            if let Ok(mut text) = sources.get_mut(grade.source) {
                text.0 = style.text;
            }
            grade.base = style.base;
            grade.glow = style.glow;
            grade.strength = style.strength;
            grade.intensity = 1.0;
            grade.pulse = 0.0;
            grade.bounce = 1.0;
        }
    }
}

/// Advances every word's fade, glow pulse, and pop, keeps it at the
/// player's configured height within the [`GradeArea`], and feeds the shader
/// its uniforms. The `x` is set once at spawn; only `y` tracks the option.
fn animate_grades(
    time: Res<Time>,
    settings: Res<PlayerSettings>,
    area: Option<Res<GradeArea>>,
    mut materials: ResMut<Assets<GradeTextMaterial>>,
    mut grades: Query<(
        &mut GradeText,
        &MeshMaterial2d<GradeTextMaterial>,
        &mut Transform,
    )>,
) {
    let delta = time.delta_secs();
    for (mut grade, material, mut transform) in &mut grades {
        if grade.intensity <= 0.0 {
            continue;
        }
        grade.intensity = (grade.intensity - delta / FADE_SECONDS).max(0.0);
        grade.pulse += delta;
        grade.bounce = (grade.bounce - delta / BOUNCE_SECONDS).max(0.0);

        if let Some(mut material) = materials.get_mut(&material.0) {
            apply_style(
                &mut material,
                grade.base,
                grade.glow,
                grade.strength,
                grade.intensity,
                glow_pulse(grade.pulse),
            );
        }

        let bounce = EaseFunction::CubicOut.sample_clamped(grade.bounce);
        transform.scale = Vec3::splat(1.0 + BOUNCE_AMOUNT * bounce);
        if let Some(area) = &area {
            transform.translation.y = grade_y(area, settings[grade.player].grade_position);
        }
    }
}

/// Publishes the play stage's [`GradeArea`] from the padded window, so grades
/// map their height option to the screen. The options preview fills its own.
fn set_stage_grade_area(config: Res<GameConfig>, windows: Query<&Window>, mut commands: Commands) {
    let Ok(window) = windows.single() else {
        return;
    };
    let half = visible_world_size(window).y / 2.0;
    let padding = config.stage.screen_edge_padding;
    commands.insert_resource(grade_area(half - padding, -half + padding));
}

/// The world Y band the grade group occupies, top (0%) to bottom (100%).
/// Each context fills it: the play stage from the padded window, the options
/// preview from the modal stripe.
#[derive(Resource, Default, Clone, Copy)]
pub struct GradeArea {
    pub top: f32,
    pub bottom: f32,
}

/// The grade word's world Y for a player's grade-position percentage within
/// its area: 0% at the top, 100% at the bottom.
pub fn grade_y(area: &GradeArea, grade_position: Percent) -> f32 {
    let t = (grade_position.0 / 100.0).clamp(0.0, 1.0);
    area.top + (area.bottom - area.top) * t
}

/// The grade area for a usable band spanning `top_edge`..`bottom_edge` (world
/// Y), inset so the word and the combo tracking under it both stay inside.
pub fn grade_area(top_edge: f32, bottom_edge: f32) -> GradeArea {
    GradeArea {
        top: top_edge - GRADE_HALF_HEIGHT,
        bottom: bottom_edge + COMBO_GAP + COMBO_HALF_HEIGHT,
    }
}

fn linear(color: Color) -> Vec3 {
    let c = color.to_linear();
    Vec3::new(c.red, c.green, c.blue)
}

/// Packs a grade's colors into the shader uniforms at a given fade and glow
/// pulse — the one place the color/glow packing lives, shared by the live
/// animation and the `render_grade` inspector.
pub fn apply_style(
    material: &mut GradeTextMaterial,
    base: Color,
    glow: Color,
    strength: f32,
    intensity: f32,
    pulse: f32,
) {
    material.base = linear(base).extend(intensity);
    material.glow = linear(glow).extend(strength * pulse);
}

/// The glow strength at a moment since the grade landed: full at the strike
/// (`seconds` 0), draining toward [`GLOW_FLOOR`].
pub fn glow_pulse(seconds: f32) -> f32 {
    GLOW_FLOOR + (1.0 - GLOW_FLOOR) * (-seconds / PULSE_TAU).exp()
}

/// The glow's drained resting level, so the word stays lit between strikes.
pub const GLOW_FLOOR_LEVEL: f32 = GLOW_FLOOR;

/// The word, base color, glow color, and glow strength one outcome shows.
pub struct GradeStyle {
    pub text: String,
    pub base: Color,
    pub glow: Color,
    pub strength: f32,
}

pub fn grade_style(config: &GameConfig, outcome: RowOutcome) -> GradeStyle {
    match outcome {
        RowOutcome::Hit { error } => {
            let Grade::Hit(grade) = config.grade(outcome) else {
                unreachable!("hits always grade into a timed grade");
            };
            let def = &config.grading.dynamic[grade.0];
            // Like ITG: the letters are white, the grade's color is the glow.
            GradeStyle {
                text: hit_text(def, error),
                base: Color::WHITE,
                glow: def.glow.color,
                strength: def.glow.strength,
            }
        }
        RowOutcome::Miss => {
            // ITG's Miss is the exception — its letters carry the red.
            let miss = &config.grading.fixed.miss;
            GradeStyle {
                text: miss.name.clone(),
                base: miss.color,
                glow: miss.glow.color,
                strength: miss.glow.strength,
            }
        }
    }
}

/// The word for a hit, marking the side of the perfect moment the input
/// fell on: early feedback leads the name, late feedback trails it.
fn hit_text(def: &DynamicGradeDef, error: Seconds) -> String {
    let name = &def.name;
    let early = error.0 > 0.0;
    // Displayed offset is input-relative: negative = early, positive = late.
    let offset_ms = (-error.to_millis()).round() as i64;
    match def.timing_feedback {
        TimingFeedback::Off => name.clone(),
        TimingFeedback::Sign if early => format!("-{name}"),
        TimingFeedback::Sign => format!("{name}-"),
        TimingFeedback::Millis if early => format!("({offset_ms}ms) {name}"),
        TimingFeedback::Millis => format!("{name} (+{offset_ms}ms)"),
    }
}

File: nodes/stepfile_player/GradeText.cs
using Godot;
using Rhythm.Core;

namespace Rhythm;

/// <summary>
/// Constants for grade text display.
/// </summary>
public static class GradeTextConstants
{
    public const float FontSize = 50.0f;
    public const float PresentW = 760.0f;
    public const float PresentH = 170.0f;
    public const float Supersample = 2.0f;
    public const float HaloRadius = 18.0f;
    public const float HaloStrength = 1.6f;
    public const float PulseTau = 0.3f;
    public const float GlowFloor = 0.32f;
    public const float GradeHalfHeight = 36.0f;
    public const float ComboHalfHeight = 24.0f;
    public const float ComboGap = 62.0f;
    public const float BounceSeconds = 0.13f;
    public const float BounceAmount = 0.18f;
    public const float FadeSeconds = 1.0f;
}

/// <summary>Bounds for grade word placement on screen.</summary>
public record GradeArea(float Top, float Bottom)
{
    public float Height => Bottom - Top;

    /// <summary>Computes y position for a grade word within this area.</summary>
    public float GradeY(Percent position)
    {
        var t = position.Value / 100.0f;
        return Top + (Height / 2.0f) * t + Height / 2.0f * (1.0f - t);
    }
}

/// <summary>
/// The pieces of one grade-text rig: the offscreen white word and the
/// shader sprite presenting it. Owners position and re-layer the sprite;
/// freeing it frees the whole rig (the viewport rides along as its child).
/// </summary>
public class GradeRig
{
    public Sprite2D Sprite { get; private set; }
    private Label _label;
    public ShaderMaterial Material { get; private set; }

    private GradeRig(Sprite2D sprite, Label label, ShaderMaterial material)
    {
        Sprite = sprite;
        _label = label;
        Material = material;
    }

    /// <summary>
    /// Builds a grade-text rig under <paramref name="layer"/>, presenting at the
    /// layer-local origin until positioned.
    /// </summary>
    public static GradeRig SpawnRig(Node2D layer)
    {
        var viewport = new SubViewport();
        viewport.TransparentBg = true;
        viewport.Size = new Vector2I(
            (int)(GradeTextConstants.PresentW * GradeTextConstants.Supersample),
            (int)(GradeTextConstants.PresentH * GradeTextConstants.Supersample)
        );

        var word = new Label();
        word.AddThemeColorOverride("font_color", Colors.White);
        var fontFile = GD.Load<FontFile>("res://assets/fonts/JetBrainsMono-Regular.ttf");
        if (fontFile != null)
        {
            word.AddThemeFontOverride("font", fontFile);
        }
        word.AddThemeFontSizeOverride("font_size", (int)(GradeTextConstants.FontSize * GradeTextConstants.Supersample));
        word.Text = "";
        viewport.AddChild(word);
        var wordSize = word.GetSize();
        word.Position = new Vector2(
            (GradeTextConstants.PresentW * GradeTextConstants.Supersample) / 2.0f - wordSize.X / 2.0f,
            (GradeTextConstants.PresentH * GradeTextConstants.Supersample) / 2.0f - wordSize.Y / 2.0f
        );

        var shader = GD.Load<Shader>("res://nodes/stepfile_player/grade_text.gdshader");
        if (shader == null)
        {
            GD.PushError("Failed to load grade_text.gdshader");
            shader = new Shader();
        }

        var material = new ShaderMaterial { Shader = shader };
        material.SetShaderParameter("shape",
            new Vector4(
                GradeTextConstants.HaloRadius / GradeTextConstants.PresentW,
                GradeTextConstants.HaloRadius / GradeTextConstants.PresentH,
                GradeTextConstants.HaloStrength,
                0.0f
            )
        );

        var sprite = new Sprite2D();
        sprite.AddChild(viewport);
        if (viewport.GetTexture() is Texture2D texture)
        {
            sprite.Texture = texture;
        }
        sprite.Scale = Vector2.One / GradeTextConstants.Supersample;
        sprite.Material = material;
        layer.AddChild(sprite);

        return new GradeRig(sprite, word, material);
    }

    /// <summary>Updates the grade word text and repositions it.</summary>
    public void SetText(string text)
    {
        _label.Text = text;
        var size = _label.GetSize();
        _label.Position = new Vector2(
            (GradeTextConstants.PresentW * GradeTextConstants.Supersample) / 2.0f - size.X / 2.0f,
            (GradeTextConstants.PresentH * GradeTextConstants.Supersample) / 2.0f - size.Y / 2.0f
        );
    }

    /// <summary>Queues the sprite for deletion.</summary>
    public void Free()
    {
        Sprite.QueueFree();
    }
}

/// <summary>
/// One player's grade word on stage, driven by the session: refreshed on
/// each graded row, animated every frame.
/// </summary>
public class GradeDisplay
{
    public PlayerId Player { get; private set; }
    public GradeRig Rig { get; private set; }
    private float _originX;
    private float _intensity;
    private float _pulse;
    private float _bounce;
    private Color _base;
    private Color _glow;
    private float _strength;

    public GradeDisplay(Node2D layer, PlayerId player, float originX)
    {
        Player = player;
        Rig = GradeRig.SpawnRig(layer);
        Rig.Sprite.Position = new Vector2(originX, 0.0f);
        Rig.Sprite.Visible = false;
        _originX = originX;
        _intensity = 0.0f;
        _pulse = 0.0f;
        _bounce = 0.0f;
        _base = Colors.White;
        _glow = Colors.White;
        _strength = 0.0f;
    }

    /// <summary>Updates the origin x position.</summary>
    public void SetOriginX(float originX)
    {
        _originX = originX;
        Rig.Sprite.Position = new Vector2(originX, 0.0f);
    }

    /// <summary>
    /// Applies a graded row outcome: updates the text, colors, and starts
    /// the bounce and glow animation.
    /// </summary>
    public void Apply(GameConfig config, RowOutcome outcome)
    {
        var grade = config.ClassifyGrade(outcome);
        var style = GradeStyleUtility.ComputeStyle(config, outcome);

        Rig.SetText(style.Text);
        _base = style.BaseColor;
        _glow = style.GlowColor;
        _strength = style.GlowStrength;
        _intensity = 1.0f;
        _pulse = 0.0f;
        _bounce = GradeTextConstants.BounceSeconds;
        Rig.Sprite.Visible = true;
    }

    /// <summary>
    /// Animates the grade word: fades intensity, pulses the glow, and bounces
    /// the sprite scale. Called every frame.
    /// </summary>
    public void Animate(float delta, float y)
    {
        _intensity = Mathf.Max(0.0f, _intensity - delta / GradeTextConstants.FadeSeconds);
        _pulse += delta;

        var glow = _strength * GlowPulse(_pulse);
        var baseColor = new Vector4(_base.R, _base.G, _base.B, _intensity);
        var glowColor = new Vector4(_glow.R, _glow.G, _glow.B, glow);

        Rig.Material.SetShaderParameter("base_color", baseColor);
        Rig.Material.SetShaderParameter("glow_color", glowColor);

        if (_bounce > 0.0f)
        {
            _bounce -= delta;
            var t = 1.0f - _bounce / GradeTextConstants.BounceSeconds;
            var easeOut = Mathf.Sin(t * Mathf.Pi / 2.0f);
            var scale = 1.0f + (1.0f - easeOut) * GradeTextConstants.BounceAmount;
            Rig.Sprite.Scale = Vector2.One * scale / GradeTextConstants.Supersample;
        }
        else if (!Rig.Sprite.Scale.IsEqualApprox(Vector2.One / GradeTextConstants.Supersample))
        {
            Rig.Sprite.Scale = Vector2.One / GradeTextConstants.Supersample;
        }

        Rig.Sprite.Position = new Vector2(_originX, y);
    }

    private static float GlowPulse(float seconds)
    {
        var tau = Mathf.Exp(-seconds / GradeTextConstants.PulseTau);
        return GradeTextConstants.GlowFloor + (1.0f - GradeTextConstants.GlowFloor) * tau;
    }
}

/// <summary>Style parameters for a grade outcome.</summary>
public record GradeStyle(string Text, Color BaseColor, Color GlowColor, float GlowStrength);

/// <summary>Utility functions for grade styling.</summary>
public static class GradeStyleUtility
{
    /// <summary>Computes the display style for a graded outcome.</summary>
    public static GradeStyle ComputeStyle(GameConfig config, RowOutcome outcome)
    {
        var grade = config.ClassifyGrade(outcome);

        return grade switch
        {
            Grade.Hit hit =>
                new GradeStyle(
                    config.Grading?.Dynamic[hit.Index.Value].Name ?? "HIT",
                    Colors.White,
                    config.Grading?.Dynamic[hit.Index.Value].GlowColor ?? Colors.Yellow,
                    config.Grading?.Dynamic[hit.Index.Value].GlowStrength ?? 1.0f
                ),
            Grade.Miss =>
                new GradeStyle(
                    config.Grading?.Miss?.Name ?? "MISS",
                    Colors.White,
                    config.Grading?.Miss?.GlowColor ?? Colors.Red,
                    config.Grading?.Miss?.GlowStrength ?? 1.0f
                ),
            _ => throw new InvalidOperationException($"Unknown grade type: {grade.GetType().Name}")
        };
    }
}
